lib/paypal_express/api.rb

```diff
@@ -83,15 +85,28 @@ def refund_payment(kb_account_id, kb_payment_id, kb_payment_transaction_id, kb_p
 
       def get_payment_info(kb_account_id, kb_payment_id, properties, context)
         t_info_plugins = super(kb_account_id, kb_payment_id, properties, context)
+        # Should never happen...
+        return [] if t_info_plugins.nil?
 
         # Completed purchases/authorizations will have two rows in the responses table (one for api_call 'build_form_descriptor', one for api_call 'purchase/authorize')
         # Other transaction types don't support the :PENDING state
         target_transaction_types = [:PURCHASE, :AUTHORIZE]
-        is_transaction_pending = t_info_plugins.find { |t_info_plugin| target_transaction_types.include?(t_info_plugin.transaction_type) && t_info_plugin.status != :PENDING }.nil?
+        only_pending_transaction = t_info_plugins.find { |t_info_plugin| target_transaction_types.include?(t_info_plugin.transaction_type) && t_info_plugin.status != :PENDING }.nil?
         t_info_plugins_without_pending = t_info_plugins.reject { |t_info_plugin| target_transaction_types.include?(t_info_plugin.transaction_type) && t_info_plugin.status == :PENDING }
 
-        # If only pending transactions exist, return the pending one. Otherwise, only return those non-pending transaction.
-        is_transaction_pending ? t_info_plugins: t_info_plugins_without_pending
+        # If its token has expired, cancel the payment and update the response row.
+        if only_pending_transaction
+          return t_info_plugins unless token_expired(t_info_plugins.last)
+          if cancel_pending_transaction(t_info_plugins.last).nil?
+            logger.warning("Try to cancel pending transaction #{t_info_plugins.last.kb_transaction_payment_id} but failed")
+            t_info_plugins
+          else
+            logger.info("Cancel pending transaction #{t_info_plugins.last.kb_transaction_payment_id}")
```

Comment:
Starting with Kill Bill 0.16.4, we switched to a Splunk-friendly message format (see http://dev.splunk.com/view/logging-best-practices/SP-CAAADP6): changing the message to `"Cancel pending kb_payment_transaction_id='#{t_info_plugins.last.kb_transaction_payment_id}'"` will let us search for `kb_payment_transaction_id` more easily.

See https://github.com/killbill/killbill-cybersource-plugin/blob/master/lib/cybersource/api.rb#L336 for example.